src/compiler/visitor.js

```diff
@@ -3,46 +3,61 @@
  * the visitor pattern so we can map AST to functions
  */
 
+let noop = function() {};
+
 let visitor = {
   build(fns) {
-    let step = function(node) {
+    let step = function(node, index, context) {
       let type = node[0];
-      let extraArgs = Array.prototype.slice.call(arguments, 1);
```

Comment:
are you not a fan of extraArgs..

this allows the developer using visitor to define the arguments

e.g.

```
let walkWithNodesOnly = visitor.build({foo(node){}, bar(node){});
let walkWithNodeAndIndex = visitor.build({foo(node, idx){}, bar(node, idx){});
let walkWithNodeAndContext = visitor.build({foo(node, ctx){}, bar(node, ctx){});

walkWithNodesOnly(ast);
walkWithNodeAndIndex(ast....)  // i guess this isn't possible
walkWithNodeAndIndex(ast, ctx);



```